go/arrow/ipc/message.go

```diff
@@ -0,0 +1,197 @@
+type Message struct {
+	refCount int64
+	msg      *flatbuf.Message
+	meta     *memory.Buffer
+	body     *memory.Buffer
+}
+
+func NewMessage(meta, body *memory.Buffer) *Message {
+	meta.Retain()
+	body.Retain()
+	return &Message{
+		refCount: 1,
+		msg:      flatbuf.GetRootAsMessage(meta.Bytes(), 0),
+		meta:     meta,
+		body:     body,
+	}
+}
+
+// Retain increases the reference count by 1.
+// Retain may be called simultaneously from multiple goroutines.
+func (msg *Message) Retain() {
+	atomic.AddInt64(&msg.refCount, 1)
+}
+
+// Release decreases the reference count by 1.
+// Release may be called simultaneously from multiple goroutines.
+// When the reference count goes to zero, the memory is freed.
+func (msg *Message) Release() {
+	debug.Assert(atomic.LoadInt64(&msg.refCount) > 0, "too many releases")
+
+	if atomic.AddInt64(&msg.refCount, -1) == 0 {
+		msg.meta.Release()
+		msg.body.Release()
+		msg.msg = nil
+		msg.meta = nil
+		msg.body = nil
+	}
+}
+
+func (msg Message) Version() MetadataVersion {
+	return MetadataVersion(msg.msg.Version())
+}
+
+func (msg Message) Type() MessageType {
+	return MessageType(msg.msg.HeaderType())
+}
+
+func (msg Message) BodyLen() int64 {
+	return msg.msg.BodyLength()
+}
+
+// MessageReader reads messages from an io.Reader.
+type MessageReader struct {
+	r io.Reader
+
+	refCount int64
+	msg      *Message
+}
+
+// NewMessageReader returns a reader that reads messages from an input stream.
+func NewMessageReader(r io.Reader) *MessageReader {
+	return &MessageReader{r: r, refCount: 1}
+}
+
+// Retain increases the reference count by 1.
+// Retain may be called simultaneously from multiple goroutines.
+func (r *MessageReader) Retain() {
+	atomic.AddInt64(&r.refCount, 1)
+}
+
+// Release decreases the reference count by 1.
+// When the reference count goes to zero, the memory is freed.
+// Release may be called simultaneously from multiple goroutines.
+func (r *MessageReader) Release() {
+	debug.Assert(atomic.LoadInt64(&r.refCount) > 0, "too many releases")
+
+	if atomic.AddInt64(&r.refCount, -1) == 0 {
+		if r.msg != nil {
+			r.msg.Release()
+			r.msg = nil
+		}
+	}
+}
+
+// Message returns the current message that has been extracted from the
+// underlying stream.
+// It is valid until the next call to Message.
+func (r *MessageReader) Message() (*Message, error) {
+	var buf = make([]byte, 4)
+	_, err := io.ReadFull(r.r, buf)
+	if err != nil {
+		return nil, errors.Wrap(err, "arrow/ipc: could not read message length")
+	}
+	msgLen := int32(binary.LittleEndian.Uint32(buf))
+	if msgLen == 0 {
+		// optional 0 EOS control message
+		return nil, io.EOF // FIXME(sbinet): send nil instead? or a special EOS error?
+	}
+
+	buf = make([]byte, msgLen)
+	_, err = io.ReadFull(r.r, buf)
+	if err != nil {
+		return nil, errors.Wrap(err, "arrow/ipc: could not read message metadata")
+	}
+	meta := memory.NewBufferBytes(buf)
+
+	bodyLen := flatbuf.GetRootAsMessage(meta.Bytes(), 0).BodyLength()
```

Comment:
We deserialize the flatbuffer message twice: on this line and with the call to NewMessage below. I don't know if it really matters...

Reply:
I'll add a TODO/FIXME.

Reply:
done. (actually dropped the double deserialization by introducing a non-exported function)